mino/minows/session.go

```diff
@@ -0,0 +1,250 @@
+	}
+	payload, err := msg.Serialize(o.rpc.context)
+	if err != nil {
+		return xerrors.Errorf("could not serialize message: %v", err)
+	}
+
+	err = encoder.Encode(&Packet{Source: src, Payload: payload})
+	if err != nil {
+		return xerrors.Errorf("could not encode packet: %v", err)
+	}
+	return nil
+}
+
+func (o orchestrator) fetch(decoder *gob.Decoder) (Packet, mino.Address,
+	error) {
+	var forward Forward
+	err := decoder.Decode(&forward)
+	if err != nil {
+		return Packet{}, nil,
+			xerrors.Errorf("could not decode packet: %v", err)
+	}
+	dest := o.rpc.mino.GetAddressFactory().
+		FromText(forward.Destination)
+	if dest == nil {
+		return Packet{}, nil,
+			xerrors.New("could not unmarshal address")
+	}
+	return forward.Packet, dest, nil
+}
+
+func (o orchestrator) relay(packet Packet, dest address) error {
+	encoder, ok := o.outs[dest.identity]
+	if !ok {
+		return xerrors.Errorf("%v: %v", ErrNotPlayer, dest)
+	}
+	err := encoder.Encode(packet)
+	if err != nil {
+		return xerrors.Errorf("could not encode packet: %v", err)
+	}
+	o.logger.Debug().Stringer("to", dest).Msgf("relayed packet")
+	return nil
+}
+
+func (o orchestrator) listen(decoder *gob.Decoder) (Packet, error) {
+	for {
+		packet, dest, err := o.fetch(decoder)
+		if err != nil {
+			return Packet{}, xerrors.Errorf("could not receive: %v", err)
+		}
+		switch to := dest.(type) {
+		case orchestratorAddr:
+			if o.myAddr.Equal(to) {
+				return packet, nil
+			}
+		case address:
+			err := o.relay(packet, to)
+			if err != nil {
+				return Packet{}, xerrors.Errorf("could not relay: %v", err)
+			}
+		}
+	}
+}
+
+type participant struct {
+	logger zerolog.Logger
+
+	myAddr address
+	rpc    rpc
+	// Connects to the orchestrator
+	out *gob.Encoder
+	in  chan Packet
+}
+
+func (p participant) Send(msg serde.Message, addrs ...mino.Address) <-chan error {
+	return doSend(addrs, msg, p.send, p.logger)
+}
+
+func (p participant) Recv(ctx context.Context) (mino.Address, serde.Message, error) {
+	unpack := unpacker(p.rpc.mino.GetAddressFactory(), p.rpc.factory, p.rpc.context)
+	return doReceive(ctx, p.in, unpack, p.logger)
+}
+
+func (p participant) send(addr mino.Address, msg serde.Message) error {
+	switch addr.(type) {
+	case address:
+	case orchestratorAddr:
+	default:
+		return xerrors.Errorf("%v: %T", ErrWrongAddressType, addr)
+	}
+
+	src, err := p.myAddr.MarshalText()
+	if err != nil {
+		return xerrors.Errorf("could not marshal address: %v", err)
+	}
+	payload, err := msg.Serialize(p.rpc.context)
+	if err != nil {
+		return xerrors.Errorf("could not serialize message: %v", err)
+	}
+	dest, err := addr.MarshalText()
+	if err != nil {
+		return xerrors.Errorf("could not marshal address: %v", err)
+	}
+
+	// Send to orchestrator to relay to the destination participant
+	forward := Forward{
+		Packet:      Packet{Source: src, Payload: payload},
+		Destination: dest,
+	}
+	err = p.out.Encode(&forward)
+	if err != nil {
+		return xerrors.Errorf("could not encode packet: %v", err)
+	}
```

Comment:
this would be so much simpler with serde, and not mixed with the `session` code.